Python/Product/Analysis/Interpreter/Ast/NameLookupContext.cs

```diff
@@ -0,0 +1,350 @@
+                } else {
+                    ctxt._scopes.Push(scope);
+                }
+            }
+
+            return ctxt;
+        }
+
+        private IPythonModule ImportBuiltinModule() {
+            var modname = Ast.LanguageVersion.Is3x() ? SharedDatabaseState.BuiltinName3x : SharedDatabaseState.BuiltinName2x;
+            var mod = Interpreter.ImportModule(modname);
+            Debug.Assert(mod != null, "Failed to import " + modname);
+            mod?.Imported(Context);
+            return mod;
+        }
+
+        public Dictionary<string, IMember> PushScope(Dictionary<string, IMember> scope = null) {
+            scope = scope ?? new Dictionary<string, IMember>();
+            _scopes.Push(scope);
+            return scope;
+        }
+
+        public Dictionary<string, IMember> PopScope() {
+            return _scopes.Pop();
+        }
+
+        internal LocationInfo GetLoc(Node node) {
+            if (!IncludeLocationInfo) {
+                return null;
+            }
+            if (node == null || node.StartIndex >= node.EndIndex) {
+                return null;
+            }
+
+            var start = node.GetStart(Ast);
+            var end = node.GetEnd(Ast);
+            return new LocationInfo(FilePath, start.Line, start.Column, end.Line, end.Column);
+        }
+
+        internal LocationInfo GetLocOfName(Node node, NameExpression header) {
+            var loc = GetLoc(node);
+            if (loc == null || header == null) {
+                return null;
+            }
+
+            var nameStart = header.GetStart(Ast);
+            if (!nameStart.IsValid) {
+                return loc;
+            }
+
+            if (nameStart.Line > loc.StartLine || (nameStart.Line == loc.StartLine && nameStart.Column > loc.StartColumn)) {
+                return new LocationInfo(loc.FilePath, nameStart.Line, nameStart.Column, loc.EndLine, loc.EndColumn);
+            }
+
+            return loc;
+        }
+
+        private string GetNameFromExpressionWorker(Expression expr) {
+            var ne = expr as NameExpression;
+            if (ne != null) {
+                return ne.Name;
+            }
+
+            var me = expr as MemberExpression;
+            if (me != null) {
+                return "{0}.{1}".FormatInvariant(GetNameFromExpressionWorker(me.Target), me.Name);
+            }
+
+            throw new FormatException();
+        }
+
+        public string GetNameFromExpression(Expression expr) {
+            try {
+                return GetNameFromExpressionWorker(expr);
+            } catch (FormatException) {
+                return null;
+            }
+        }
+
+        public IMember GetValueFromExpression(Expression expr) {
+            return GetValueFromExpression(expr, DefaultLookupOptions);
+        }
+
+        public IMember GetValueFromExpression(Expression expr, LookupOptions options) {
+            var ne = expr as NameExpression;
```

Comment:
I agree with both of you. I'm on board with using `if (expr is NameExpression ne)`, and I'm not convinced regarding the new switch features.